add arize.md (#475)
algorithmiaio/dev-center: _pages/integrations/arize.md

{% highlight python %}
algorithmia>=1.0.0,<2.0
joblib==1.0.1
numpy>=1.0
pandas<2.0,>=0.25.3
scikit-learn<1.0
shap==0.39.0
{% endhighlight %}

Train your model, generate some predictions, and then serialize the trained model:

{% highlight python %}
import os

import Algorithmia
import joblib
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

# Load and split data.
from sklearn import datasets
data = datasets.load_breast_cancer()
X, y = datasets.load_breast_cancer(return_X_y=True)
X, y = (
    pd.DataFrame(X.astype(np.float32), columns=data["feature_names"]),
    pd.Series(y)
)

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

# Fit a simple logistic regression model.
clf = LogisticRegression(max_iter=3000, verbose=False).fit(X_train, y_train)

# Generate predictions from the model.
y_train_pred = clf.predict(X_train)
y_test_pred = clf.predict(X_test)

# Create a directory to save serialized model files.
model_dir = "models/"
if not os.path.exists(model_dir):
    os.mkdir(model_dir)

# Serialize the model.

> model_file = "Algorithmia_Tutorial_Model.joblib"

Reviewer: I wouldn't mind if we added here versioning the model file - not a dealbreaker, but I think it would be nice to start showing that in our examples since it implicitly calls out that we don't automatically version data files

model_file_path = model_dir+model_file
joblib.dump(clf, model_file_path)
{% endhighlight %}

## Uploading your trained model to Algorithmia

> Use the following code to upload your model to a hosted data collection on Algorithmia, without ever leaving your training envronemnt. Note that if you're running [Algorithmia Enterprise](/enterprise), you'll need to specify the API endpoint `CLUSTER_DOMAIN` when you create the Algorithmia `client` object; if not, delete the references to this variable.

Reviewer: ```suggestion
Use the following code to upload your model to a hosted data collection on Algorithmia, without ever leaving your training environment. Note that if you're running [Algorithmia Enterprise](/enterprise), you'll need to specify the API endpoint `CLUSTER_DOMAIN` when you create the Algorithmia `client` object; if not, delete the references to this variable.
```

Author: lol wow really blew it on the spelling there

> You'll need to replace the `COLLECTION_OWNER` string with the name of the user or org account that owns the collection to which you'll be uploading your model and from which your algorithm will then load that model on the Algorithmia side, and the `COLLECTION_NAME` string with the name of that collection. The Algorithmia API key you're using must have write access to this data collection. See our [Hosted Data](/developers/data/hosted) docs for information about how to use hosted data collections.

Reviewer: ```suggestion
You'll need to replace the `COLLECTION_OWNER` string with the name of the user or org account that owns the data collection. You will upload your model to that data collection and in your algorithm source code you will replace the `COLLECTION_NAME` string with the name of that data collection. The Algorithmia API key you're using must have write access to this data collection. See our [Hosted Data](/developers/data/hosted) docs for information about how to use hosted data collections.
```

shap_values = get_shap_values(clf, X, show_graph=True)
{% endhighlight %}

<img src="{{site.cdnurl}}{{site.baseurl}}/images/post_images/arize/shap_values.png" alt="Shap values plot in Jupyter notebook" class="screenshot">

The following represents the algorithm code that you would deploy on Algorithma, not the training platform used above. {:.notice-info}

## Setting up your Algorithmia environment
To begin, on Algorithmia you'll need to [create an algorithm](/developers/algorithm-development/your-first-algo) using an environment with Python 3.6 or later. In your algorithm's `requirements.txt` file, add the Arize Python library to add the monitoring capabilities provided by Arize, as well as the `shap` library and the standard ML dependencies `joblib` and `pandas`:

{% highlight python %}
algorithmia>=1.0.0,<2.0
arize>=2.1.1
joblib==1.0.1
pandas<2.0,>=0.25.3
shap==0.39.0
{% endhighlight %}

## Deploying your model on Algorithmia
Set the environment variables `ARIZE_API_KEY` and `ARIZE_ORG_KEY` with your Arize keys; these keys; these secrets are accessible through the [Arize Settings page](https://app.arize.com/admin).

> Remember from above that you must also set the `ALGORITHMIA_API_KEY` environment variable with the value of your Algorithmia API key if you're running the algorithm from outside of the Algorithmia Web IDE; this API key will data-read permissions. You must also replace the `COLLECTION_OWNER` and `COLLECTION_NAME` strings with the account name and collection name where the model is stored. 

Reviewer: ```suggestion
Remember from above that you must also set the `ALGORITHMIA_API_KEY` environment variable with the value of your Algorithmia API key if you're running the algorithm from outside of the Algorithmia Web IDE; this API key will have read access only. You must also replace the `COLLECTION_OWNER` and `COLLECTION_NAME` strings with the account name and collection name where the model is stored. 
```

Author: changed to say "only needs to have" instead of "will have"

    shap_values = pd.DataFrame(shap_values, columns=X_data.columns)

    ids = pd.Series([str(uuid.uuid4()) for _ in range(len(X_data))])

    # Log the prediction data to Arize.
    log_responses = arize_client.log_bulk_predictions(
        model_id="Algorithmia_Tutorial_Model",
        model_version="1.0",
        model_type=ModelTypes.BINARY,
        features=X_data,
        prediction_ids=ids,
        prediction_labels=pd.Series(y_pred))

    # Log the SHAP data to Arize.
    shap_responses = arize_client.log_bulk_shap_values(
        model_id="Algorithmia_Tutorial_Model",
        # Pass in the same prediction IDs to match predictions and actuals.
        prediction_ids=ids,
        shap_values=shap_values
    )

    # Return prediction data.
    res = pd.DataFrame(y_pred)
    res.index = ids
    res.index.rename("pred_id", inplace=True)
    return res.to_json()
{% endhighlight %}

The following code is intended to be executed back in the same external environment (Jupyter notebook or external training platform) that you used above to train your algorithm, once you've built the algorithm on Algorithmia. {:.notice-info}

> Once you've built your algorithm, you can call it using its version hash to test it out; this will be a value like `f35025657bdc37eb0d6ffeed62b0539ee21c8b4e`. If you build your algorithm in the browser IDE, this hash is displayed in the test console output upon successful build completion, but it's also available in the "Builds" tab on the algorithm's homepage. You can also publish the algorithm, in which case you'll be able to call the algorithm using a semantic version such as `1.0.0`.

Reviewer: ```suggestion
Once you've built your algorithm, you can call it using its version hash to test it out; this will be a value like `f35025657bdc37eb0d6ffeed62b0539ee21c8b4e`. If you build your algorithm in the Algorithmia Web IDE, this hash is displayed in the test console output upon successful build completion, but it's also available in the "Builds" tab on the algorithm's homepage. You can also publish the algorithm, in which case you'll be able to call the algorithm using a semantic version such as `1.0.0`.
```

In the code below, substitute the appropriate strings for `ALGO_OWNER` (the user or org account under which the algorithm was created), `ALGO_NAME` (the name of the algorithm), and `ALGO_VERSION` (the version hash or semantic version described above). As in the code above when you originally uploaded your model to Algorithmia, the `CLUSTER_DOMAIN` variable should be deleted if you aren't using an Enterprise cluster.

{% highlight python %}
import Algorithmia

# Define variables to uniquely identify algorithm.
ALGO_OWNER = "ALGO_OWNER"
ALGO_NAME = "ALGO_NAME"
ALGO_VERSION = "ALGO_VERSION"

# Build algorithm identifier and instantiate client.
ALGO_IDENTIFIER = ALGO_OWNER+"/"+ALGO_NAME+"/"+ALGO_VERSION
client = Algorithmia.client(ALGORITHMIA_API_KEY, CLUSTER_DOMAIN)
algo = client.algo(ALGO_IDENTIFIER)

> # Optionally set timeout parameters for testing purposes.

Reviewer: Either link to details/more info about timeout params or add more info here in comments

Author: added link.
